tun/tun_illumos.go

```diff
@@ -0,0 +1,355 @@
+//go:build illumos && amd64
+// +build illumos,amd64
+
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright (C) 2017-2021 WireGuard LLC. All Rights Reserved.
+ * Copyright 2019 Joyent, Inc.
+ */
+
+package tun
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+const (
+	tunNode = "/dev/tun"
+)
+
+type NativeTun struct {
+	fd      int
+	name    string
+	tunFile *os.File
+	events  chan Event
+	errors  chan error
+}
+
+// XXX This is largely copied from the freebsd code...
+// XXX How do I test if it is actually working?
+func (tun *NativeTun) routineRouteListener() {
+	var (
+		statusUp  bool
+		statusMTU int
+	)
+
+	tunIfindex, err := func() (int, error) {
```

Comment:
I know you copied this from the freebsd implementation, but it doesn't make much sense to me to define an in-place closure and execute it once at the time of definition. More idiomatic would be to have this either as a separate method on NativeTun, so this callsite turns into `tun.index()`. Or, since it's only used once, remove the surrounding closure and make it straight-line code that does the InterfaceByName lookup followed by extracting the index. The resulting code would be shorter too, since you'd have a single error check instead of a pair.